dartagnan/src/main/java/com/dat3m/dartagnan/expression/processing/ExprSimplifier.java

```diff
@@ -62,54 +63,34 @@ public Expression visit(Atom atom) {
     }
 
     @Override
-    public BExpr visit(BConst bConst) {
-        return bConst;
+    public Expression visit(BExprBin bBin) {
+        Expression left = bBin.getLHS().visit(this);
+        Expression right = bBin.getRHS().visit(this);
+        if (left instanceof BConst constant) {
+            boolean value = constant.getValue();
+            return switch (bBin.getOp()) {
+                case OR -> value ? left : right;
+                case AND -> value ? right : left;
+            };
+        }
+        if (right instanceof BConst constant) {
+            boolean value = constant.getValue();
+            return switch (bBin.getOp()) {
+                case OR -> value ? right : left;
+                case AND -> value ? left : right;
+            };
+        }
+        return expressions.makeBinary(left, bBin.getOp(), right);
     }
 
     @Override
-    public BExpr visit(BExprBin bBin) {
-        // Due to constant propagation we are not guaranteed to get BExprs
-        if(!(bBin.getLHS().visit(this) instanceof BExpr && bBin.getRHS().visit(this) instanceof BExpr)) {
-            return bBin;
-        }
-        BExpr lhs = (BExpr) bBin.getLHS().visit(this);
-        BExpr rhs = (BExpr) bBin.getRHS().visit(this);
-        switch (bBin.getOp()) {
-            case OR:
-                if (lhs.isTrue() || rhs.isTrue()) {
-                    return BConst.TRUE;
-                } else if (lhs.isFalse()) {
-                    return rhs;
-                } else if (rhs.isFalse()) {
-                    return lhs;
-                }
-                break;
-            case AND:
-                if (lhs.isFalse() || rhs.isFalse()) {
-                    return BConst.FALSE;
-                } else if (lhs.isTrue()) {
-                    return rhs;
-                } else if (rhs.isTrue()) {
-                    return lhs;
-                }
-                break;
-        }
-        return expressions.makeBinary(lhs, bBin.getOp(), rhs);
-    }
-
-    @Override
-    public BExpr visit(BExprUn bUn) {
-        // Due to constant propagation we are not guaranteed to get BExprs
-        Expression innerExpr = bUn.getInner().visit(this);
-        if(!(innerExpr instanceof BExpr)) {
-            return bUn;
-        }
-        BExpr inner = (BExpr) innerExpr;
-        if (inner instanceof BConst) {
-            return expressions.makeValue(!inner.isTrue());
+    public Expression visit(BExprUn bUn) {
+        Expression inner = bUn.getInner().visit(this);
+        if (inner instanceof BConst constant) {
+            return expressions.makeValue(!constant.getValue());
```

Comment:
Maybe assert that the unary operator is indeed negation just to be sure that we do not accidentally break this in the future.